Customize presentation of accept/decline buttons in iMip mail

brad2014 opened the pull request:
- Fix Issue #11230
Only present accept/decline button links in iMip mail for REQUEST, not CANCEL or others.

- Fix Issue #12156
Implement config setting "dav.invitation_link_recipients", to control
which invitation recipients see accept/decline button links.  The
default, for public internet facing servers, is to always include
them.  For a server on a private intranet, this setting can be set
to the email addresses or email domains of users whose browsers can
access the nextcloud server referenced by those accept/decline
button links. It can also be set to "false" to exclude the links
from all requests.

Signed-off-by: Brad Rubenstein <brad@wbr.tech>

Commit:
Customize presentation of accept/decline buttons in iMip mail

Fix Issue #11230
Only present accept/decline button links in iMip mail for REQUEST, not CANCEL or others.

Fix Issue #12156
Implement config setting "dav.invitation_link_recipients", to control
which invitation recipients see accept/decline button links.  The
default, for public internet facing servers, is to always include
them.  For a server on a private intranet, this setting can be set
to the email addresses or email domains of users whose browsers can
access the nextcloud server referenced by those accept/decline
button links. It can also be set to "false" to exclude the links
from all requests.

Signed-off-by: Brad Rubenstein <brad@wbr.tech>

Commit:
corrected sample config

Signed-off-by: Brad Rubenstein <brad@wbr.tech>

brad2014 commented:
Yes, you are correct, thanks!  Fixed.

brad2014 commented on apps/dav/lib/CalDAV/Schedule/IMipPlugin.php:
```diff
@@ -144,11 +144,11 @@ public function schedule(Message $iTipMessage) {
 		$summary = $iTipMessage->message->VEVENT->SUMMARY;
 
-		if (parse_url($iTipMessage->sender, PHP_URL_SCHEME) !== 'mailto') {
+		if (strcasecmp(parse_url($iTipMessage->sender, PHP_URL_SCHEME), 'mailto') !== 0) {
```
Note: One more place where mailto scheme comparison should be case insensitive (if the event contained MAILTO, then it would not have properly been detected).  This is related to the fix nextcloud/calendar#8afe26b3 by @georgehrke.

brad2014 commented:
Another note:  It occurs to me that RFC5545 defaults ATTENDEE rsvpparm to RSVP=FALSE.  Perhaps the invite buttons should only be presented when (a) the attendee has a mailto, and (b) the method is REQUEST, and (c) RSVP=TRUE.  (The examples in the RFC suggest it would be set false for optional attendees and non-individuals, e.g. resources). The code does not currently look at the RSVP parameter, as far as I can tell.  May I ask @georgehrke or @MorrisJobke for advice? I'm happy to put it in if it seems the right thing to do.

brad2014 commented:
> 
> Most of this is not done in Nextcloud directly, but in Sabre/Dav.
> But skimming over https://github.com/nextcloud/3rdparty/blob/214c4155f587f5178d792fe4a839044bdc9982f1/sabre/vobject/lib/ITip/Broker.php this doesn't seem to be the case indeed.

My notes: The sabre ITip code properly passes the client's RSVP parameter through unchanged, except when processing an incoming reply (to remove the RSVP=TRUE parameter when an RSVP has been received).  Other than that, it is up to the client to decide when RSVP's are desired (e.g. apps/calendar (at the UI) or server/dav (by nextcloud policy) chooses whether rsvp's are desired from "required" or "optional" participants, and, on the other end, the server/dav/lib/CalDAV/Schedule/IMipPlugin.php must decide what to present in emails when the organizer (or nextcloud by policy) has set RSVP to true or false when they created they event that generates the REQUEST.

brad2014 commented:
> Thx for your pull-request!
> Can you please add unit-tests for your changes?

I need to learn how to use phpUnit. :-) 

I'll look for a doc that tells me how to set up my local machine to run nextcloud tests (pointers welcome).

Commit:
Respect RSVP parameter for attendees when adding accept/decline buttons.

If RSVP=TRUE parameter is FALSE or absent for an ATTENDEE, then do no
present accept/decline buttons. The organizer isn't asking for an RSVP.

Signed-off-by: Brad Rubenstein <brad@wbr.tech>

Commit:
Revert 3ff3ed0c56 case-insensitive compares.

My oops.  The comparisons, which are copied from the IMipPlugin shipped with sabre-io/dav,
do not need to be case insensitive because the sender and recipient names are normalized by sabre,
(see calls to getNormalizedValue in voboject/lib/ITip/Broker.php).

Signed-off-by: Brad Rubenstein <brad@wbr.tech>

brad2014 commented:
@georgehrke or @MorrisJobke : I'm looking at the existing tests, to find a model for a new unit test, and I think I need some guidance.  I'm not sure where to ask, so I'll ask here.  I hope you don't mind.

1. I read [unit testing](https://docs.nextcloud.com/server/15/developer_manual/core/unit-testing.html).  I thought I'd first try to get testing working.  On a clean Fedora-28 VM with the [nextcloud dependencies](https://docs.nextcloud.com/server/stable/admin_manual/installation/source_installation.html#prerequisites-for-manual-installation) installed, and the [latest phpunit](https://phpunit.readthedocs.io/en/8.0/installation.html).  But when I run ./autotest.sh, I get

```
$ ./autotest.sh
Using PHP executable /usr/bin/php
PHP Warning:  require_once(/vagrant/server/3rdparty/autoload.php): failed to open stream: No such file or directory in /vagrant/server/build/OCPSinceChecker.php on line 23
PHP Fatal error:  require_once(): Failed opening required '/vagrant/server/3rdparty/autoload.php' (include_path='.:/usr/share/pear:/usr/share/php') in /vagrant/server/build/OCPSinceChecker.php on line 23
```

What am I missing?

2. Any suggestions on how to make a unit test for this or where to put it in the tree?  Are there any existing caldav tests I can use for a model?

MorrisJobke commented:
> What am I missing?

Did you ran `git submodule update --init` to initialize the 3rdparty git submodule?

brad2014 commented:
Thanks.

> Did you ran `git submodule update --init` to initialize the 3rdparty git submodule?

Thanks.  Not sure how I missed it - the docs are clear.

brad2014 commented:
>     1. Any suggestions on how to make a unit test for this or where to put it in the tree?  Are there any existing caldav tests I can use for a model?

I think I've figured it out.  I will write some unit tests shortly.

Commit:
IMipPlugin: Added unit test for dav.invitation_link_recipients support

Existing tests required modification to correctly mock up the new config
parameter fetch, and to set the RSVP flag for attendees (since the test
was detecting token generation, and we no longer generate tokens when
no RSVP is requested by the client or sent by nextcloud).

Signed-off-by: Brad Rubenstein <brad@wbr.tech>

Commit:
IMipPlugin: DRY up - move common test set-up into setUp function

Signed-off-by: Brad Rubenstein <brad@wbr.tech>

brad2014 commented:
@georgehrke, I have added unit tests per your suggestion - In the server-generated iMIP email, I'm using link token generation as a surrogate for whether buttons and links are added (the test doesn't actually look at the generated HTML - perhaps it should). 

I also did the following:

1. Rebased my changes (trivially) to the top of the master branch.
2. Moved all the repeated in apps/dav/tests/unit/CalDAV/Schedule/IMipPluginTest.php into a phpUnit setUp function.  The minimal to the test file is 1dbda26, and the refactor follows in 1a29239.  (It is 43% DRYer)

The tests pass, I should add, pass :smile:.

brad2014 commented:
[This is my first pull request for nextcloud - if you all have any suggestions or concerns about my code or process, I'm very interested in hearing them.]

MorrisJobke commented:
Hi @brad2014 - sorry that we didn't had time to look into this for the Nextcloud 16 milestone. We were quite busy with other tasks. We still appreciate the work you put into this, but the freeze for Nextcloud 16 is active since last Friday and I will put this into the Nextcloud 17 bucket. I hope that is okay for you.

brad2014 commented:
> Hi @brad2014 - sorry that we didn't had time to look into this for the Nextcloud 16 milestone. We were quite busy with other tasks. We still appreciate the work you put into this, but the freeze for Nextcloud 16 is active since last Friday and I will put this into the Nextcloud 17 bucket. I hope that is okay for you.

No worries - thanks for the update.

brad2014 commented:
@georgehrke @MorrisJobke I'd like to do a few more PR's, but I'd also like to wait until this first one is pulled (if you think it's pullable).  May I get some feedback on the test I wrote?  Does it run as part of the continuous-integration drone?  (I looked at the drone failure, but it seems unrelated to my change, yes?)

skjnldsv commented:
Small comment to say that your pr looks pretty awesome! Nicely written, tests... You rocks! :star2: 
:hugs:

blizzz commented:
We'll be able to merge once stable16 is branched off.

georgehrke commented on apps/dav/lib/CalDAV/Schedule/IMipPlugin.php:
```diff
@@ -239,9 +239,44 @@ public function schedule(Message $iTipMessage) {
+			** to the email address or email domain, or array of addresses or domains,
+			** of recipients who can access the server.
+			**
+			** To deliver URL's always, set invitation_link_recipients to boolean "true".
```
To always deliver URLs

georgehrke replied:
Related to my request to change it to `getAppValue`: We usually use "yes" and "no"

brad2014 replied:
fixed in 0de8c54

georgehrke commented on apps/dav/lib/CalDAV/Schedule/IMipPlugin.php:
```diff
@@ -239,9 +239,44 @@ public function schedule(Message $iTipMessage) {
+			** of recipients who can access the server.
+			**
+			** To deliver URL's always, set invitation_link_recipients to boolean "true".
+			** To suppress URL's entirely, set invitation_link_recipients to boolean "false".
```
To suppress URLs entirely

brad2014 replied:
fixed in 0de85c4

brad2014 commented on apps/dav/lib/CalDAV/Schedule/IMipPlugin.php:
```diff
@@ -239,9 +239,44 @@ public function schedule(Message $iTipMessage) {
+			*/
+
+			$recipientDomain = substr(strrchr($recipient, "@"), 1);
+			$invitationLinkRecipients = $this->config->getSystemValue('dav.invitation_link_recipients', true);
```
I'm not clear on where (a) the documentation would go to describe the setting if it is the DB (that corresponds to the current PR's doc in config.sample.php), and (b) how a system admin would set the value of the setting.  Would I need to create an administrative UI to do that?  Is the expectation that they would manually add a row to the oc_appconfig table?

If I go this route, I'd change the variable type to simply be a string 'yes', 'no', or a comma separated list of domains.  (no boolean/string/array stuff).  I agree that is cleaner.

brad2014 replied:
Fixed in 0de8c54.  There is no "UI" to set or add the value in this PR, but if the value is in the DB, it will properly be used to include or suppress imip invitation accept/decline links.

brad2014 committed:
Move dav.invitation_link_recipients from getSystemValue to getAppValue

Per @georgehrke change request for PR #12392, instead of setting
dav.invitation_link_recipients in the system config.php file, we
set it in the database table oc_appconfig.

Furthermore, the value of the config variable is always a string:
'yes' to include links in imip mail, 'no' to exclude them, or a
comma-separated list of email addresses and/or domains for which
they should be included.  If not specified in oc_appconfig, the
default is 'yes'.

Signed-off-by: brad2014 <brad2014@users.noreply.github.com>

brad2014 commented on apps/dav/lib/CalDAV/Schedule/IMipPlugin.php:
```diff
@@ -249,28 +249,26 @@ public function schedule(Message $iTipMessage) {
-				 || (is_string($invitationLinkRecipients) && strcasecmp($recipientDomain, $invitationLinkRecipients) === 0)
-				 || (is_array($invitationLinkRecipients) && in_array(strtolower($recipient), $invitationLinkRecipients))
-				 || (is_array($invitationLinkRecipients) && in_array(strtolower($recipientDomain), $invitationLinkRecipients))) {
+			$invitationLinkRecipients = explode(',', preg_replace('/\s+/', '', strtolower($this->config->getAppValue('dav', 'invitation_link_recipients', 'yes'))));
```
I am normalizing the AppConfig value by lowercasing it and trimming spaces to make comparisons more reliable.  I also notice I'm assuming explode always returns an array (no error cases); I think that's right.

brad2014 commented:
@georgehrke  I have addressed all your review comments on the branch, but I'm not sure if the next step is for me to do something to remove the "Changes Requested" blocker, or if that's something you do after additional review. Please advise. Thank you!

georgehrke commented:
@brad2014 The changes requested blocker will go after another review. It's already late in Germany, but i will look into this first thing tomorrow morning. :)

brad2014 commented:
Thank you (didn't mean to rush anyone, just wasn't sure if I needed to do anything).

georgehrke reviewed (approved):
👍 

Let me check why CI didn't start

georgehrke commented:
https://drone.nextcloud.com/nextcloud/server/20578

georgehrke commented:
CI is happy: https://drone.nextcloud.com/nextcloud/server/20598

georgehrke commented:
please review @rullzer @juliushaertl @ChristophWurst :)
